File: src/RehabEasy.Infrastructure/Services/PatientRecordHelper.cs
using System.Text.Json;
using RehabEasy.Domain.Models;

namespace RehabEasy.Infrastructure.Services;

public static class PatientRecordHelper
{
    public const string TestTypeCvTug = "CvTUG";
    public const string TestTypeEquilibrio = "Equilibrio";
    public const string TestTypeOther = "Outro";

    public static string? TryGetPatientExternalId(string rawPayloadJson)
    {
        return TryGetPatientString(rawPayloadJson, "external_id");
    }

    public static string? TryGetPatientName(string rawPayloadJson)
    {
        return TryGetPatientString(rawPayloadJson, "name");
    }

    public static string ResolveTestType(string rawPayloadJson)
    {
        if (string.IsNullOrWhiteSpace(rawPayloadJson))
        {
            return TestTypeOther;
        }

        try
        {
            using JsonDocument document = JsonDocument.Parse(rawPayloadJson);
            JsonElement root = document.RootElement;

            string? explicitType = TryGetString(root, "test_type")
                ?? TryGetNestedString(root, "assessment", "test_type")
                ?? TryGetNestedString(root, "assessment", "type");

            if (!string.IsNullOrWhiteSpace(explicitType))
            {
                return NormalizeTestType(explicitType);
            }

            string? sender = TryGetString(root, "sender")
                ?? TryGetString(root, "source")
                ?? TryGetString(root, "source_system");

            if (string.Equals(sender, TestTypeCvTug, StringComparison.OrdinalIgnoreCase))
            {
                return TestTypeCvTug;
            }

            if (string.Equals(sender, "Posturografia VR", StringComparison.OrdinalIgnoreCase))
            {
                return TestTypeEquilibrio;
            }

            if (TryGetProperty(root, "assessment", out JsonElement assessment))
            {
                if (TryGetProperty(assessment, "posturographic_indices", out _))
                {
                    return TestTypeEquilibrio;
                }

                if (TryGetProperty(assessment, "conditions", out _) &&
                    TryGetProperty(root, "patient", out _))
                {
                    return TestTypeCvTug;
                }
            }

            return TestTypeOther;
        }
        catch (JsonException)
        {
            return TestTypeOther;
        }
    }

    public static string ResolvePatientId(RehabEasyRecord record)
    {
        if (!string.IsNullOrWhiteSpace(record.PatientId))
        {
            return record.PatientId.Trim();
        }

        return TryGetPatientExternalId(record.RawPayloadJson)?.Trim() ?? string.Empty;
    }

    public static string ResolveTestType(RehabEasyRecord record)
    {
        if (!string.IsNullOrWhiteSpace(record.TestType))
        {
            return record.TestType;
        }

        return ResolveTestType(record.RawPayloadJson);
    }

    private static string NormalizeTestType(string testType)
    {
        if (testType.Contains("tug", StringComparison.OrdinalIgnoreCase))
        {
            return TestTypeCvTug;
        }

        if (testType.Contains("equil", StringComparison.OrdinalIgnoreCase) ||
            testType.Contains("posturo", StringComparison.OrdinalIgnoreCase))
        {
            return TestTypeEquilibrio;
        }

        return testType.Trim();
    }

    private static string? TryGetPatientString(string rawPayloadJson, string propertyName)
    {
        if (string.IsNullOrWhiteSpace(rawPayloadJson))
        {
            return null;
        }

        try
        {
            using JsonDocument document = JsonDocument.Parse(rawPayloadJson);
            JsonElement root = document.RootElement;

            if (!TryGetProperty(root, "patient", out JsonElement patient))
            {
                return null;
            }

            return TryGetString(patient, propertyName);
        }
        catch (JsonException)
        {
            return null;
        }
    }

    private static string? TryGetNestedString(JsonElement root, string objectName, string propertyName)
    {
        return TryGetProperty(root, objectName, out JsonElement nested)
            ? TryGetString(nested, propertyName)
            : null;
    }

    private static bool TryGetProperty(JsonElement element, string propertyName, out JsonElement value)
    {
        foreach (JsonProperty property in element.EnumerateObject())
        {
            if (string.Equals(property.Name, propertyName, StringComparison.OrdinalIgnoreCase))
            {
                value = property.Value;
                return true;
            }
        }

        value = default;
        return false;
    }

    private static string? TryGetString(JsonElement element, string propertyName)
    {
        return TryGetProperty(element, propertyName, out JsonElement value) &&
               value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;
    }
}

File: java/src/main/java/com/rehabeasy/service/PatientRecordHelper.java
package com.rehabeasy.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.rehabeasy.json.JsonSupport;
import com.rehabeasy.model.RehabEasyRecord;

public final class PatientRecordHelper {
    public static final String TEST_TYPE_CVTUG = "CvTUG";
    public static final String TEST_TYPE_EQUILIBRIO = "Equilibrio";
    public static final String TEST_TYPE_INDEX_INDEX = "Index-Index";
    public static final String TEST_TYPE_OTHER = "Outro";

    private PatientRecordHelper() {
    }

    public static String tryGetPatientExternalId(String rawPayloadJson) {
        return tryGetPatientString(rawPayloadJson, "external_id");
    }

    public static String tryGetPatientName(String rawPayloadJson) {
        return tryGetPatientString(rawPayloadJson, "name");
    }

    public static String resolveTestType(String rawPayloadJson) {
        if (rawPayloadJson == null || rawPayloadJson.isBlank()) {
            return TEST_TYPE_OTHER;
        }

        try {
            JsonNode root = JsonSupport.readTree(rawPayloadJson);
            String explicitType = firstNonBlank(
                    JsonSupport.string(root, "test_type"),
                    JsonSupport.string(JsonSupport.property(root, "assessment"), "test_type"),
                    JsonSupport.string(JsonSupport.property(root, "assessment"), "type"));
            if (explicitType != null) {
                return normalizeTestType(explicitType);
            }

            String sender = firstNonBlank(
                    JsonSupport.string(root, "sender"),
                    JsonSupport.string(root, "source"),
                    JsonSupport.string(root, "source_system"));
            if (equalsIgnoreCase(sender, TEST_TYPE_CVTUG)) {
                return TEST_TYPE_CVTUG;
            }
            if (equalsIgnoreCase(sender, "Posturografia VR")) {
                return TEST_TYPE_EQUILIBRIO;
            }
            if (equalsIgnoreCase(sender, TEST_TYPE_INDEX_INDEX) || equalsIgnoreCase(sender, "index-index")) {
                return TEST_TYPE_INDEX_INDEX;
            }

            JsonNode assessment = JsonSupport.property(root, "assessment");
            String assessmentTestType = JsonSupport.string(assessment, "test_type");
            if (assessmentTestType != null && assessmentTestType.toLowerCase().contains("index")) {
                return TEST_TYPE_INDEX_INDEX;
            }
            JsonNode metrics = JsonSupport.property(assessment, "metrics");
            if (!JsonSupport.property(metrics, "final_fingertip_distance_mm").isMissingNode()
                    && !JsonSupport.property(metrics, "final_fingertip_distance_mm").isNull()) {
                return TEST_TYPE_INDEX_INDEX;
            }
            if (!JsonSupport.property(assessment, "posturographic_indices").isMissingNode()
                    && !JsonSupport.property(assessment, "posturographic_indices").isNull()) {
                return TEST_TYPE_EQUILIBRIO;
            }
            if (!JsonSupport.property(assessment, "conditions").isMissingNode()
                    && !JsonSupport.property(assessment, "conditions").isNull()
                    && !JsonSupport.property(root, "patient").isMissingNode()
                    && !JsonSupport.property(root, "patient").isNull()) {
                return TEST_TYPE_CVTUG;
            }
            return TEST_TYPE_OTHER;
        } catch (IllegalArgumentException ignored) {
            return TEST_TYPE_OTHER;
        }
    }

    public static String resolvePatientId(RehabEasyRecord record) {
        if (record.patientId() != null && !record.patientId().isBlank()) {
            return record.patientId().trim();
        }
        String externalId = tryGetPatientExternalId(record.rawPayloadJson());
        return externalId == null ? "" : externalId.trim();
    }

    public static String resolveTestType(RehabEasyRecord record) {
        return record.testType() != null && !record.testType().isBlank()
                ? record.testType()
                : resolveTestType(record.rawPayloadJson());
    }

    private static String normalizeTestType(String testType) {
        String normalized = testType.toLowerCase();
        if (normalized.contains("tug")) {
            return TEST_TYPE_CVTUG;
        }
        if (normalized.contains("equil") || normalized.contains("posturo")) {
            return TEST_TYPE_EQUILIBRIO;
        }
        if (normalized.contains("index")) {
            return TEST_TYPE_INDEX_INDEX;
        }
        return testType.trim();
    }

    private static String tryGetPatientString(String rawPayloadJson, String propertyName) {
        if (rawPayloadJson == null || rawPayloadJson.isBlank()) {
            return null;
        }
        try {
            JsonNode patient = JsonSupport.property(JsonSupport.readTree(rawPayloadJson), "patient");
            return JsonSupport.string(patient, propertyName);
        } catch (IllegalArgumentException ignored) {
            return null;
        }
    }

    private static String firstNonBlank(String... values) {
        for (String value : values) {
            if (value != null && !value.isBlank()) {
                return value;
            }
        }
        return null;
    }

    private static boolean equalsIgnoreCase(String left, String right) {
        return left != null && left.equalsIgnoreCase(right);
    }
}
